### PR TITLE
Expose sqlite's sqlite3_create_collation()

### PR DESCRIPTION
**Pull request by TaKO8Ki**
I've taken over https://github.com/diesel-rs/diesel/pull/2495.

### CODE REVIEW

File: diesel/src/sqlite/connection/raw.rs [hunk 1]
```diff
@@ -292,99 +343,155 @@ extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A
...
+    };
+
+    // It doesn't matter if f is UnwindSafe, since we abort on panic.
+    let f = std::panic::AssertUnwindSafe(|| match f(rhs, lhs) {
+        std::cmp::Ordering::Greater => 1,
+        std::cmp::Ordering::Equal => 0,
+        std::cmp::Ordering::Less => -1,
+    });
+    let result = std::panic::catch_unwind(f);
+    result.unwrap_or_else(|_| std::process::abort())
```
Comment (weiznich) [inline diesel/src/sqlite/connection/raw.rs:491]: We should just set the an error here like done for the other both cases above.

**Review by weiznich — COMMENTED** [review]
Thanks for reactivating this PR I left another round of comments. As this is touching unsafe code I would like to get this correct.
I encourage @diesel-rs/reviewers to have an independent look at the unsafe code parts.

File: diesel/src/sqlite/connection/raw.rs (hunk 1 above)
Comment (TaKO8Ki) [reply diesel/src/sqlite/connection/raw.rs:491]: So you mean I should set an error by using `context_error_str`?

File: diesel/src/sqlite/connection/raw.rs [hunk 2]
```diff
@@ -204,7 +251,8 @@ extern "C" fn run_custom_function<F>(
 ) where
     F: FnMut(&RawConnection, &[*mut ffi::sqlite3_value]) -> QueryResult<SerializedValue>
         + Send
-        + 'static,
+        + 'static
+        + std::panic::RefUnwindSafe,
 {
```
[the remark this answers is not in the document]
Comment (TaKO8Ki) [reply diesel/src/sqlite/connection/raw.rs:256]: I've fixed this!

File: diesel/src/sqlite/connection/raw.rs (hunk 1 above)
Comment (weiznich) [reply diesel/src/sqlite/connection/raw.rs:489]: :+1:

**Comment by TaKO8Ki** [comment]
Thank you for your help. Looks good to me.

File: CHANGELOG.md [hunk 3]
```diff
@@ -203,6 +205,8 @@ for Rust libraries in [RFC #1105](https://github.com/rust-lang/rfcs/blob/master/
 
 * Queries containing a `distinct on` clause check now on compile time that a compatible order clause was set.
 
+* Implementations of custom SQLite SQL functions now probably check for panics
```
Comment (weiznich) [inline CHANGELOG.md:208]:
```suggestion
* Implementations of custom SQLite SQL functions now check for panics
```

File: diesel/src/sqlite/connection/functions.rs [hunk 4]
```diff
@@ -37,6 +37,7 @@ where
 
         process_sql_function_result::<RetSqlType, Ret>(result)
     })?;
+
```
Comment (weiznich) [inline diesel/src/sqlite/connection/functions.rs:40]:
```suggestion
```

File: diesel/src/sqlite/connection/raw.rs [hunk 5]
```diff
@@ -196,57 +245,96 @@ fn convert_to_string_and_free(err_msg: *const libc::c_char) -> String {
...
+}
+
+struct CustomFunctionUserPtr<F> {
+    callback: F,
+    function_name: String,
+}
+
 #[allow(warnings)]
-extern "C" fn run_custom_function<F>(
+extern "C" fn run_custom_function<'b, F>(
```
Comment (weiznich) [inline diesel/src/sqlite/connection/raw.rs:283]:
```suggestion
extern "C" fn run_custom_function<F>(
```

**Pull request merged**